_ap/19/index.adoc

```diff
@@ -1,78 +1,58 @@
+title: "Toward a more unified UX: BFF"
 status: "Draft"
 authors:
-- "andreaTP"
+  - "Andrea Peruffo"
+  - "Riccardo Forina"
+  - "Eric Wittmann"
 tags: []
 ---
 
-// Top style tips:
-// * Use one sentence per line
-// * No unexpanded acronyms
-// * No undefined jargon
-
-// You don't have to use the following sections, but they provide a 
-// useful structure for writing a clear document.
-
 ## Intent
-// Summarize in a single sentence what the pattern tries to achieve.
-//
-// For example:
-//
-// Define a policy for exposing metrics to customers in a consistent way.
+
+Define an architectural component to better decouple backend services and frontend needs offering a more cohesive UX across devices and tools (such as UI and CLI).
 
 ## Motivation
 
-// In a few paragraphs describe the motivating factors for this pattern
-//
-// For example:
-// 
-// Prometheus scrape endpoints are the defacto standard for exposing metric information for collection.
-// However this format and the associated conventions are not sufficient in themselves to ensure
-// that metrics are exposed in a consistent way. Problems include:
-// * Insufficient guidance on metric and label naming to provide consistency and 
-//   establishing the principal of hiding implementation details.
-// * Treatment of this as a first-class API, with established mechanisms for API evolution
-// * etc.
+UI and CLI needs to poll information from multiple separate sources causing:
+
+* poor UX, there is need for multiple network calls to offer simple functionalities
+* code duplication, data aggregation logic is scattered and duplicated across different users
+* testing and mocking, decoupling frontend and backend systems with only one component
 
 ## Applicability
 
-// Call out when this pattern might apply (and when it should not apply, if relevant)
-// For example:
-//
-// This pattern applies to any service which exposes metrics as Prometheus-scrape endpoints.
-// It does not apply to services which provide access to metrics in other formats, or via
-// means other than an API (e.g. metric visualization via a system such as Grafana).
+This pattern applies when the functionality offered by frontend/CLI/etc. needs to access various APIs and do data-aggregation logic in various forms.
+In those cases placing a frontend opinionated component in front can be an appropriate solution to speed up development and be a single source of truth for integration logic.
 
 ## Description
 
-// Describe the pattern in enough detail that a fresh pair of eyes can understand it
-// and apply it elsewhere.
-//
-// For this metrics example we might link to existing guidelines on naming, augment them with extra rules 
-// which apply to services, explain how the API would allow for graceful API evolution without "flag days".
+Implementing a https://samnewman.io/patterns/architectural/bff/[BFF] for the backend services to better decouple and centralize the integration logic.
```

Comment:
Add the `propagate the access token` concept in the description.
Mention the alternative being GraphQL but that for the team, expertise was not deems high enough to explore that path. This will leave the doors for other teams to explore.

Reply:
1. done
2. I do believe that the current wording reflect weel enough our position toward GraphQL: https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/99/commits/877e4c8dac5be9ea3b2736bb0c73004adea04531

Reply:
On 2, it's in the ADR and not in the AP, It is worth repeating to help consumption of the materials.

Reply:
Any reference to the implementing technology has been deliberately omitted from the AP, unless you have a really strong opinion I would keep it this way.

Reply:
Yes I think it's important. It's not an implementation detail to go for a custom REST API vs a GraphQL based API, this is a defining choice.

Reply:
> this is a defining choice

can you please elaborate?
This interpretation is completely alien to me and, I believe, to anyone else involved in drafting this proposal.
cc. @EricWittmann @riccardo-forina

Reply:
I'm also for not mentioning any particular technology here. A BFF could be implemented in a number of ways and there is no "best" or "worst" technology, but a "right" one that fully depends on the context.

For example, for a team that's using Typescript for the full stack and fully control the clients connecting to the backend, tRPC is a great choice, arguably better than GraphQL or plain old REST+OpenAPI

Reply:
@emmanuelbernard if you are convinced that this aspect is important for the AP I'm happy to remove the AP from this PR and let you(or someone else) take over in another iteration.
In this case, I disagree with mixing scopes:
- BFF is the Architectural Pattern we want to promote as a possible solution to common issues
- OpenAPI/GraphQL is the technical instrument used to implement it, which is a completely orthogonal discussion - i.e. it can be implemented with CORBA or any other technology that serves the purpose

We briefly touched on this point with @chirino and I felt alignment on this aspect.

Reply:
Lol you mention COBRA in the same breath as GraphQL as being fitting for a BFF pattern.  I’m not sure you understood our conversation at all.

Reply:
I am proposing changes as part of a new commit. See https://github.com/andreaTP/architecture/pull/1

Reply:
Thanks @emmanuelbernard , the context you proposed is fine for me, merged 👍 

@chirino I firmly believe that CORBA is a perfectly valid fit for an application in COBOL ( [ref](http://computer-programming-forum.com/48-cobol/f4fdc89fc2d9bfb3.htm) )